### PR TITLE
[GEOS-11338] CapabilityUtil SearchMinMaxScaleDenominator should include support for multiple NamedLayers

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOS-11338](https://osgeo-org.atlassian.net/browse/GEOS-11338) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Review by prushforth — APPROVED** [review]
Looks good to me!

File: src/main/src/main/java/org/geoserver/catalog/impl/StyleInfoImpl.java [hunk 1]
```diff
@@ -133,6 +133,10 @@ public Style getStyle() throws IOException {
 
     @Override
     public StyledLayerDescriptor getSLD() throws IOException {
+        // to avoid test NPEs
+        if (catalog == null || catalog.getResourcePool() == null) {
+            return null;
+        }
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/impl/StyleInfoImpl.java:139]: Is this needed for tests? Like, why didn't we get a NPE before now? Layer group defined by styles use this method in other places in the code base.

File: src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java [hunk 2]
```diff
@@ -43,36 +48,68 @@ public static NumberRange<Double> searchMinMaxScaleDenominator(Set<StyleInfo> st
             throws IOException {
         // searches the maximum and minimum denominator in the style's rules that are contained in
         // the style set.
-        double minScaleDenominator = Double.POSITIVE_INFINITY;
-        double maxScaleDenominator = Double.NEGATIVE_INFINITY;
+        MinMaxDenominator minMaxDenominator = new MinMaxDenominator();
+        minMaxDenominator.setMinScaleDenominator(Double.POSITIVE_INFINITY);
+        minMaxDenominator.setMaxScaleDenominator(Double.NEGATIVE_INFINITY);
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java:53]: Sounds like the state you'd want right out of the constructor, without having to use setters from outside.

File: src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java [hunk 3]
```diff
@@ -231,4 +268,46 @@ public static String getGroupDefaultStyleName(String groupName) {
     public static String getGroupDefaultStyleName(LayerGroupInfo groupName) {
         return getGroupDefaultStyleName(groupName.prefixedName());
     }
+
+    /** Stores the Min and Max scale denominators for a set of styles. */
+    private static class MinMaxDenominator {
+        private Double minScaleDenominator;
+        private Double maxScaleDenominator;
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java:275]: Maybe just call them min and max for brevity?

File: src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java [hunk 4]
```diff
@@ -43,36 +48,68 @@ public static NumberRange<Double> searchMinMaxScaleDenominator(Set<StyleInfo> st
...
         for (StyleInfo styleInfo : styles) {
-            Style style = styleInfo.getStyle();
-            for (FeatureTypeStyle fts : style.featureTypeStyles()) {
-
-                for (Rule rule : fts.rules()) {
-
-                    if (rule.getMinScaleDenominator() < minScaleDenominator) {
-                        minScaleDenominator = rule.getMinScaleDenominator();
+            if (styleInfo.getSLD() != null && styleInfo.getSLD().getStyledLayers() != null) {
+                for (StyledLayer styledLayer : styleInfo.getSLD().getStyledLayers()) {
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java:57]: The path towards ``styleInfo.getSLD().getStyledLayers()``  is traversed multiple times. Maybe:

```java
Optional<List<StyledLayer>> styledLayers = Optional.ofNullable(styleInfo.getSLD()).map(sld -> sld.getStyledLayers());
if (styledLayers.isPresent()) {
   for (StyledLayer styledLayer : styledLayers.get()) {
   ...
   }
} else if (styleInfo.getStyle() != null) {
   ...
}
```

**Review by aaime — CHANGES_REQUESTED** [review]
Overall looks good.

The files in ``src/extension/wps-jdbc/...`` should not be part of the commit.
Maybe add them to gitignore while you're at it.

Some comments follow below.

File: src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java (hunk 2 above)
Comment (turingtestfail) [reply src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java]: Done

File: src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java (hunk 3 above)
Comment (turingtestfail) [reply src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java]: Done

File: src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java (hunk 4 above)
Comment (turingtestfail) [reply src/wms/src/main/java/org/geoserver/wms/capabilities/CapabilityUtil.java]: Done

**Comment by aaime** [comment]
Looks good to me, feel free to merge when the github actions are green.

**Review by aaime — APPROVED** [review]

**Pull request merged**